### DISCUSSION TITLE
In Nested Components @this.set() in JS through this.components.componentsById[componentId] is undefined

### DISCUSSION THREAD

**Discussion by BilawalDeveloper** [livewire/livewire#2284] [1. Help]
Hi there,
  I have three level nested components.

**Paraent**
  `<div>
        <livewire:exams.datesheet.datesheets-list :key="'exam_datesheet_list'" />
    </div>`

**Child-A**
`<livewire:exams.datesheet.syllabus-list :datesheetId="$viewDatesheet" :key="'exam_datesheet_syllabus_list'"/>`

**Child-B inside Child-A**
`<livewire:exams.datesheet.syllabus-list.edit :syllabusId="$editExamSyllabusId" :datesheetId="$datesheetId" :key="'edit_single_subject_syllabus'"/>`

First time everything work fine but when i come back form **Child-B to Child-A** and then go back
to **Child-B** it through error this.components.componentsById[componentId] is undefined

**Code in Child-B**

`window.addEventListener('create-syllabus-js', event => {`
`initializeTinyMacyEditor("syllabus", function (editor) {`
        `editor.on("init change", function () {`
           ` editor.save();`
        `});`
        `editor.on("change", function () {`
            `@this.set("syllabu", editor.getContent());`
       ` });`
    `});`
  `});`

![Screenshot_1](https://user-images.githubusercontent.com/45271305/103425393-e3628780-4bd3-11eb-9c70-4de5089eefb9.png)
![Screenshot_2](https://user-images.githubusercontent.com/45271305/103425395-e52c4b00-4bd3-11eb-8c82-4b8392ebd07b.png)

**Comment by joshhanley** [comment]
@BilawalDeveloper can you inspect your code in devtools in your browser and try and find where `@this.set("syllabu", editor.getContent());` is and double check that `@this` is being resolved to a component ID correctly?

The other thing to check, is to make sure LivewireScripts loads first before any other scripts and libraries, so it should be something like this at the bottom of your body tag

```html
    <livewire:scripts /> <!-- or @livewireScripts-->
    <script src="some/other/libraries/like/tinymce.js"></scripts>
    <script>//Custom JS stuff here</script>
    @stack('scripts')
</body>
```

Hope this helps!

**Reply by BilawalDeveloper** [reply]
@joshhanley I try to load LivewireScripts before any scripts in my application but no luck.

I'm using livewire in existing application, is that is a problem?

**Reply by BilawalDeveloper** [reply]
@joshhanley One more question can we use **Livewire** without **Jetstream**?

**Reply by joshhanley** [reply]
@BilawalDeveloper Livewire can be used in existing applications, normally, without any issues. 

Due you have VueJS running in the app though? That can cause some strange things to happen. But there is a Livewire VueJs adapter if you do.

Definitely! Livewire has been around before Jetstream even existed, Jetstream just uses Livewire as one of its options. But you don't need Jetstream to run Livewire 🙂

**Reply by BilawalDeveloper** [reply]
@joshhanley no i'm not using VueJS.

![Screenshot_9](https://user-images.githubusercontent.com/45271305/103704440-21441f00-4fcb-11eb-9859-292b53a6d990.png)
![Screenshot_8](https://user-images.githubusercontent.com/45271305/103704471-328d2b80-4fcb-11eb-860b-95b80d39fe26.png)
![Screenshot_10](https://user-images.githubusercontent.com/45271305/103704489-391ba300-4fcb-11eb-9126-7b3b0a914337.png)
![Screenshot_11](https://user-images.githubusercontent.com/45271305/103704505-3faa1a80-4fcb-11eb-9f54-9766573b3031.png)

**Reply by joshhanley** [reply]
@BilawalDeveloper that looks like a different error than you had before. 

To fix this error in your latest screenshot `Livewire.on is not a function error` you will need to move the `Livewire.on` call inside your `livewire:load` closure above it so it runs after Livewire has loaded, because currently it is running before, hence the error.

![Screen Shot 2021-01-06 at 9 34 46 am](https://user-images.githubusercontent.com/882837/103707524-de05a280-5002-11eb-85d6-0646b659f329.png)
